### PR TITLE
Refactor sparse tensor code

### PR DESCRIPTION
**Pull request by reminisce**
This PR refactors the sparse tensor code according to the CR comments in this https://github.com/dmlc/mxnet/pull/5800.

### CODE REVIEW

**Comment by eric-haibin-lin** [comment]
Could you fix window build?
```
         c:\projects\mxnet\src\common\../operator/nn/cast_storage-inl.h(70): error C3861: 'accumulate': identifier not found (C:\projects\mxnet\src\operator\upsampling.cc) [C:\projects\mxnet\build\mxnet.vcxproj]

```

File: include/mxnet/ndarray.h [hunk 1]
```diff
@@ -767,6 +744,14 @@ class NDArray {
...
   }
 
+  /*!
+   * \brief Slice a NDArray with non-default storage
+   * Do not call this function directly. Use NDArray::Slice()
+   * \param begin begin index in first dim (inclusive)
+   * \param end end index in first dim (exclusive)
+   * \return sliced NDArray
+   */
+  NDArray SliceEx(index_t begin, index_t end) const;
```
Comment (eric-haibin-lin) [inline include/mxnet/ndarray.h]: I think it's okay to add an `if` statement to the existing `Slice` function

File: src/c_api/c_api.cc [hunk 2]
```diff
@@ -320,6 +320,7 @@ int MXNDArraySlice(NDArrayHandle handle,
   API_END_HANDLE_ERROR(delete ptr);
 }
 
+#if 0
```
Comment (eric-haibin-lin) [inline src/c_api/c_api.cc]: Just remove it?

File: src/ndarray/ndarray.cc [hunk 3]
```diff
@@ -380,6 +391,105 @@ void ScalarOp(const NDArray &lhs,
...
+  // Allocate storage
+  to->CheckAndAllocAuxData(csr::kIndPtr, from.aux_shape(csr::kIndPtr));
+  to->CheckAndAllocAuxData(csr::kIdx, from.aux_shape(csr::kIdx));
+  to->CheckAndAllocData(from.aux_shape(csr::kIdx));
+  // FIXME This is a naive implementation for CSR copy. It, however, is
+  // not efficient when the source CSR is sliced. In that case, we're copying
+  // a superset of values and indices of the slice.
+  // Ideally, we should truncate the values and indices array, and adjust indptr
+  // accordingly.
+  TBlob val = to->data();
```
Comment (eric-haibin-lin) [inline src/ndarray/ndarray.cc]: These comments can be removed

File: src/operator/nn/cast_storage-inl.h [hunk 4]
```diff
@@ -0,0 +1,300 @@
...
+#if MXNET_USE_CUDA
+inline void CastStorageDnsRspImpl(mshadow::Stream<gpu>* s, const TBlob& dns, NDArray* rsp) {
+  LOG(FATAL) << "CastStorageDnsRspImpl gpu version is not implemented.";
+}
+
+inline void CastStorageDnsCsrImpl(mshadow::Stream<gpu>* s, const TBlob& dns, NDArray* csr) {
+  LOG(FATAL) << "CastStorageDnsCsrImpl gpu version is not implemented.";
+}
+#endif
+
```
Comment (eric-haibin-lin) [inline src/operator/nn/cast_storage-inl.h]: can we move it to a `cuh` file now?

File: src/operator/nn/cast_storage-inl.h [hunk 5]
```diff
@@ -0,0 +1,300 @@
+/*!
+ *  Copyright (c) 2017 by Contributors
+ * \file cast_storage-inl.h
```
Comment (eric-haibin-lin) [inline src/operator/nn/cast_storage-inl.h]: why is this file under /nn? would it be more appropriate to be under /tensor?

**Comment by eric-haibin-lin** [comment]
Awesome work!!! Thanks for refactoring the code. One step closer to the merge

**Comment by reminisce** [comment]
The lint is clean. I will add "#include<numeric>" to fix Windows build.

File: include/mxnet/ndarray.h (hunk 1 above)
Comment (reminisce) [inline include/mxnet/ndarray.h]: Ok. I will merge SliceEx into Slice and delete SliceEx.

File: src/c_api/c_api.cc (hunk 2 above)
Comment (reminisce) [inline src/c_api/c_api.cc]: Right, I forgot to do so.

File: src/ndarray/ndarray.cc (hunk 3 above)
Comment (reminisce) [inline src/ndarray/ndarray.cc]: Ok.

File: src/operator/nn/cast_storage-inl.h (hunk 5 above)
Comment (reminisce) [inline src/operator/nn/cast_storage-inl.h]: src/operator/nn is the folder where all MXNet ops will be moved to eventually. You can see utility functions for convolution and pooling have already been moved there.

File: src/operator/nn/cast_storage-inl.h (hunk 4 above)
Comment (reminisce) [inline src/operator/nn/cast_storage-inl.h]: Sure.

**Comment by eric-haibin-lin** [comment]
just to double check, did you test all the gpu tests?

**Comment by eric-haibin-lin** [comment]
Some gpu tests failed after the refactor: http://ec2-52-25-96-65.us-west-2.compute.amazonaws.com/blue/organizations/jenkins/mxnet/detail/PR-5800/85/pipeline

**Pull request merged**